persist/seeder.go

```diff
@@ -0,0 +1,160 @@
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-kbucket"
+	inet "github.com/libp2p/go-libp2p-net"
+)
+
+// SeedDialGracePeriod is the grace period for one dial attempt
+var SeedDialGracePeriod = 5 * time.Second
+
+// TotalSeedDialGracePeriod is the total grace period for a group of dial attempts
+var TotalSeedDialGracePeriod = 30 * time.Second
+
+// NSimultaneousDial is the number of peers we will dial simultaneously
+var NSimultaneousDial = 50
+
+var ErrPartialSeed = errors.New("routing table seeded partially")
+
+var DefaultRndSeederTarget = 35
+
+type randomSeeder struct {
+	host   host.Host
+	target int
+}
+
+var _ Seeder = (*randomSeeder)(nil)
+
+// NewRandomSeeder returns a Seeder that seeds a routing table with `target` random peers from
+// the supplied candidate set, resorting to fallback peers if the candidates are unworkable.
+//
+// The fallback peers are guaranteed to exist in the peerstore.
+func NewRandomSeeder(host host.Host, target int) Seeder {
+	return &randomSeeder{host, target}
+}
+
+func (rs *randomSeeder) Seed(into *kbucket.RoutingTable, candidates []peer.ID, fallback []peer.ID) error {
+	// copy the candidates & shuffle to randomize seeding
+	cpy := make([]peer.ID, len(candidates))
+	copy(cpy, candidates)
+	rand.Shuffle(len(cpy), func(i, j int) {
+		cpy[i], cpy[j] = cpy[j], cpy[i]
+	})
+
+	left := rs.target
+	addPeer := func(p peer.ID) { // adds a peer to the routing table and decrements the left counter if successful.
+		evicted, err := into.Update(p)
+		if err == nil && evicted == "" {
+			left-- // if we evict a peer, do not decrement the counter.
+		} else if err != nil {
+			logSeed.Warningf("error while adding candidate to routing table: %s", err)
+		}
+	}
+
+	// make a list of the candidates we need to dial to to ensure they are available
+	var todial []peer.ID
+	for _, p := range cpy {
+		if left == 0 {
+			return nil // lucky case: we were already connected to all our candidates.
+		}
+		if rs.host.Network().Connectedness(p) == inet.Connected {
+			// Note: initial seeding MUST be done before dht registers for connection notifications
+			// this is because once we register for notifs, peers will get added to the RT upon connection itself
+			// we don't want that as it will wrongly mark the peer as more active in the k-bucket & move it to the front of the bucket
+			// take a look at notif.go
+			// the long term solution to this problem is being discussed at https://github.com/libp2p/go-libp2p-kad-dht/issues/283
+			addPeer(p)
+			continue
+		}
+		if addrs := rs.host.Peerstore().Addrs(p); len(addrs) == 0 {
+			logSeed.Infof("discarding routing table candidate as we no longer have addresses: %s", p)
+			continue
+		}
+		todial = append(todial, p)
+	}
+
+	type result struct {
+		p   peer.ID
+		err error
+	}
+
+	// attempts to dial to a given peer to verify it's available
+	dialFn := func(ctx context.Context, p peer.ID, res chan<- result) {
```

Comment:
I'm having second thoughts about putting this dialing logic here. It's probable that all seeders would want to verify they can _actually_ add a peer before doing so. The dialing logic needs to be somewhere common.

I wonder if the seeder should be dumb, poll-driven and not perform any dials. We can move the dialing logic to the DHT, and pass in the current state of the routing table for the seeder to use when making decisions about which candidates to return next.

```go
type Seeder interface {
   // Next returns a slice of peers to attempt to add next, based on the current
   // state of the routing table. Returning a nil slice signals that the seeder
   // is done proposing peers.
   Next(rt *kbucket.RoutingTable) []peer.ID
}
```

Reply:
@raulk  

I don't see the value in making the seeder poll driven & moving the burden of actually seeding the RT on the caller. Can you explain why that'd be a good thing ? Also, why should we remove the candidate/fallback peers from the interface ? The set of fallback peers is already configurable which means the caller can pass in peers obtained from any source.

However, I do agree that the dialing logic is pretty generic & can be pulled out. We can put stuff like that in a helpers package that makes it easy for users to construct their own seeder implementations.

Reply:
In a nutshell, I'm worried about putting dialling functionality in the Seeder. It simply doesn't belong there and it complicates the seeder, which is just supposed to seed, not dial. I see this being a future footgun, as it's essentially leaking an abstraction and turning into "spooky action at a distance". I did have it like this in my WIP work, so I totally understand that you followed suit, but that was mostly a placeholder -- I was supposed to revisit that aspect at some point.

Solutions I can think of:
1. The Seeder acts like a "proposer", and we call it iteratively until we exhaust the candidates, and the Seeder returns no more proposals. Imagine we have peers [A-E] available.
    * The Seeder implements `Next(rt *RoutingTable, candidates []peer.ID, fallback []peer.ID) (proposed []peer.ID)`.
    * On a first iteration, we call `Next()` with an empty routing table, all candidates ([A-E]), and the fallback peers.
    * The Seeder returns A and B.
    * The DHT bootstrap logic verifies that A works, B is undiallable. It adds A to the routing table, and drops B from the candidate set.
    * We call `Next()` with the routing table with A, and candidates [C-E] (having dropped B).
    * This goes on until the candidate set is exhausted, and the dialer returns nil.
2. Pass in a "validate function" to the seeder, owned by the DHT. This func would perform the dial. The Seeder calls this function and can therefore verify if the peer is diallable without owning the logic.

I still prefer 1. WDYT?

Reply:
@raulk  I implemented **1** to get a feel for what you were saying & it is a good idea indeed. Now that we have decoupled 'proposing' seeders from 'seeding', users can simply plugin custom proposing strategies (preferring candidates in ascending order of XOR distance from self for example) & the seeding will out of the box. Thanks for this !

Reply:
@raulk I implemented 1 to get a feel for what you were saying & it is a good idea indeed. Now that we have decoupled 'proposing' seeders from 'dialling' & 'seeding' the RT, users can simply plug in custom 'proposal' strategies (one strategy could be preferring candidates in ascending order of XOR distance from self for example) & the 'dialling'/'seeding' will just work out of the box. Thanks for this !